### PR TITLE
Recipe for X-UMX (using MUSDB18)

### PR DESCRIPTION
**Pull request by r-sawata**
Thank @TE-StefanUhlich, @faroit and @mpariente for lots of efforts!! I would like to do PR regarding X-UMX. 

The differences between `official master` and `xumx_branch` are summarized as follows:
[newly added]
  ･`asteroid/tests/models/xumx_test.py`
  ･`asteroid/egs/musdb18/X-UMX`
  ･`asteroid/asteroid/models/x_umx.py`
[modified]
  ･`asteroid/asteroid/models/__init__.py`
  ･`asteroid/asteroid/utils/hub_utils.py`

### CODE REVIEW

**Comment by mpariente** [comment]
Thanks @r-sawata !

Just for the background, this is going to be one of the baselines of the Music demixing challenge and was subject to heavy review in a private fork by @faroit, @TE-StefanUhlich and me.

I'll make a last review tomorrow.

File: egs/musdb18/X-UMX/run.sh [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+eval_use_gpu=-1
+
+. utils/parse_options.sh
+
+if [[ $stage -le  0 ]]; then
+  echo "Stage 0: Downloading MUSDB18 into $musdb18_dir"
+  wget -c --tries=0 --read-timeout=20 https://zenodo.org/record/1117372/files/musdb18.zip -P $musdb18_dir
+  mkdir -p $musdb18_dir/logs
+  unzip $musdb18_dir/musdb18.zip -d $musdb18_dir >> $musdb18_dir/logs/unzip_musdb18.log
+  musdbconvert $musdb18_dir $musdb18_dir
```
Comment (mpariente) [inline egs/musdb18/X-UMX/run.sh:55]: Did you install musdbconvert?

File: egs/musdb18/X-UMX/local/conf.yml [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+optim:
+  optimizer: adam
+  lr: 0.001
+  patience: 1000
+  lr_decay_patience: 80
+  lr_decay_gamma: 0.3
+  weight_decay: 0.00001
+# Data config
+data:
+  train_dir: # Please insert here the path to the MUSDB18 dataset
```
Comment (mpariente) [inline egs/musdb18/X-UMX/local/conf.yml:19]: This is automatically set in the `run.sh` right? If yes, remove that comment, it's confusing.

File: tests/models/xumx_test.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+
+def test_model_loading():
+    sources_tmp = ["bass", "drums", "vocals", "other"]
+    model = XUMX(sources=sources_tmp)
+
+    model_conf = model.serialize()
+
+    new_model = XUMX.from_pretrained(model_conf)
+    assert model.in_chan == new_model.in_chan
```
Comment (mpariente) [inline tests/models/xumx_test.py:77]: Rather test that the forwards on a random signals are the same. 
Loading and having the same `in_chan` doesn't mean much.

File: egs/musdb18/X-UMX/train.py [hunk 4]
```diff
@@ -0,0 +1,481 @@
...
+            If you select ``loss_use_multidomain: yes'' via conf.yml, this is set as True.
+        mix_coef (float): A mixing parameter for multi domain losses
+
+    References
+        [1] "All for One and One for All: Improving Music Separation by Bridging
+        Networks", Ryosuke Sawata, Stefan Uhlich, Shusuke Takahashi and Yuki Mitsufuji.
+        https://arxiv.org/abs/2010.04228 (and ICASSP 2021)
+    """
+
+    def __init__(self, args):
```
Comment (mpariente) [inline egs/musdb18/X-UMX/train.py:237]: Receiving `args` in a class is bad practice. 
You have to read the code to know what it uses, it's not transparent. 

Let's not change it to release faster, but next time, let's avoid this.

File: tests/models/xumx_test.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        spec_power=spec_power,
+        return_time_signals=return_time_signals,
+    )
+    x_umx = x_umx.eval()
+    with torch.no_grad():
+        x_umx(data)
+
+
+def test_get_model_args():
+    sources_tmp = {"src_0": "vocals"}
```
Comment (mpariente) [inline tests/models/xumx_test.py:49]: Should this be a dict or a list?

**Review by mpariente — APPROVED** [review]
I made few changes, and asked a few small others. 
When those are addressed, we can merge it.

File: egs/musdb18/X-UMX/run.sh (hunk 1 above)
Comment (r-sawata) [reply egs/musdb18/X-UMX/run.sh:55]: No, it was one of the function of `musdb`.

Namely, we can use `musdbconvert` after following:
`pip install musdb`

File: egs/musdb18/X-UMX/local/conf.yml (hunk 2 above)
Comment (r-sawata) [reply egs/musdb18/X-UMX/local/conf.yml:19]: Yes, unless running from `run.sh`.
OK, I will remove this part.

File: tests/models/xumx_test.py (hunk 3 above)
Comment (r-sawata) [reply tests/models/xumx_test.py:77]: OK, thank you for your pointing our. I will remove this `assert`.

File: egs/musdb18/X-UMX/train.py (hunk 4 above)
Comment (r-sawata) [reply egs/musdb18/X-UMX/train.py:237]: OK, I understand.
I'm sorry for my inappropriate implementation, and appreciate your advise for the next time!

File: tests/models/xumx_test.py (hunk 3 above)
Comment (mpariente) [reply tests/models/xumx_test.py:77]: But you should test the consistency in forward instead.

File: egs/musdb18/X-UMX/run.sh (hunk 1 above)
Comment (mpariente) [reply egs/musdb18/X-UMX/run.sh:55]: So you should add this deps in the requirements.txt , and install the requirements.txt in run.sh

File: tests/models/xumx_test.py (hunk 5 above)
Comment (r-sawata) [reply tests/models/xumx_test.py:49]: Sorry for my careless mistake! This part is old one, and should be type of list.

File: egs/musdb18/X-UMX/run.sh (hunk 1 above)
Comment (r-sawata) [reply egs/musdb18/X-UMX/run.sh:55]: OK, I understand and follow it.

File: tests/models/xumx_test.py (hunk 5 above)
Comment (r-sawata) [reply tests/models/xumx_test.py]: Resolved by #490 fa6ae66

File: egs/musdb18/X-UMX/local/conf.yml (hunk 2 above)
Comment (r-sawata) [reply egs/musdb18/X-UMX/local/conf.yml]: Resolved by #490 fa6ae66

File: egs/musdb18/X-UMX/run.sh (hunk 1 above)
Comment (r-sawata) [reply egs/musdb18/X-UMX/run.sh:55]: Resolved by #490 fa6ae66

File: tests/models/xumx_test.py (hunk 3 above)
Comment (r-sawata) [reply tests/models/xumx_test.py]: At #490 fa6ae66, I changed this part as follows:
```
def test_model_loading():
    sources_tmp = ["bass", "drums", "vocals", "other"]
    model = XUMX(sources=sources_tmp)

    model_conf = model.serialize()

    new_model = XUMX.from_pretrained(model_conf)

    random_input = torch.rand(3, 2, 44100, requires_grad=False)
    model = model.eval()
    new_model = new_model.eval()
    with torch.no_grad():
        output1 = model(random_input)
        output2 = new_model(random_input)
    assert torch.allclose(output1[0], output2[0])
```
Furthermore, I confirmed that there is no errors in this part.
Does this convince you?

Comment (mpariente) [reply tests/models/xumx_test.py]: This seems good, I'll have a last look before merging! 
Thanks

File: egs/musdb18/X-UMX/local/conf.yml [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+optim:
+  optimizer: adam
+  lr: 0.001
+  patience: 1000
+  lr_decay_patience: 80
+  lr_decay_gamma: 0.3
+  weight_decay: 0.00001
+# Data config
+data:
+  train_dir: /data
```
Comment (mpariente) [inline egs/musdb18/X-UMX/local/conf.yml:19]:
```suggestion
  train_dir: ./data
```

File: egs/musdb18/X-UMX/run.sh [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+mix_coef=10.0
+val_dur=80.0
+
+# Evaluation
+eval_use_gpu=-1
+
+. utils/parse_options.sh
+
+if [[ $stage -le  0 ]]; then
+  echo "Stage 0: Downloading MUSDB18 into $musdb18_dir"
```
Comment (mpariente) [inline egs/musdb18/X-UMX/run.sh:51]: Requirements should be installed at some point.

```suggestion
  echo "Stage 0: Downloading MUSDB18 into $musdb18_dir"
  pip install -r requirements.txt
```

**Review by mpariente — COMMENTED** [review]
Can you confirm about the requirements installation that it's ok for you please?
We can merge afterwards

**Comment by r-sawata** [comment]
Thank you for your revising, @mpariente! I'm fine for all of your suggestions.

**Comment by mpariente** [comment]
Merged! Thanks to all involved !

**Comment by r-sawata** [comment]
@mpariente 
Sorry! I forgot adding 2 packages: `museval` and `norbert`.
Could you add followings to `egs/musdb18/X-UMX/requirements.txt` in master branch?
```
museval>=0.4.0
norbert>=0.2.1
```

I'm sorry for your inconvenient.

**Comment by mpariente** [comment]
I let you make a new PR

**Comment by r-sawata** [comment]
OK, I will follow your suggestion. Thanks!

**Pull request merged**